### PR TITLE
fix(raft): resend configuration to members that lost their configuration

### LINKED ISSUE #15381: NullPointerException when join is retried after a restart
While testing the impact of pod restarts during dynamic scaling, we observed the following error and partition join was stuck. 
```
java.lang.NullPointerException: Cannot invoke "io.atomix.raft.storage.system.Configuration.index()" because the return value of "io.atomix.raft.cluster.impl.RaftClusterContext.getConfiguration()" is null
	at io.atomix.raft.impl.RaftContext.setCommitIndex(RaftContext.java:503) ~[zeebe-atomix-cluster-8.4.0-SNAPSHOT.jar:8.4.0-SNAPSHOT]
	at io.atomix.raft.roles.PassiveRole.appendEntries(PassiveRole.java:588) ~[zeebe-atomix-cluster-8.4.0-SNAPSHOT.jar:8.4.0-SNAPSHOT]
	at io.atomix.raft.roles.PassiveRole.handleAppend(PassiveRole.java:426) ~[zeebe-atomix-cluster-8.4.0-SNAPSHOT.jar:8.4.0-SNAPSHOT]
	at io.atomix.raft.roles.PassiveRole.onAppend(PassiveRole.java:349) ~[zeebe-atomix-cluster-8.4.0-SNAPSHOT.jar:8.4.0-SNAPSHOT]
	at io.atomix.raft.impl.RaftContext.lambda$registerHandlers$19(RaftContext.java:323) ~[zeebe-atomix-cluster-8.4.0-SNAPSHOT.jar:8.4.0-SNAPSHOT]
	at io.atomix.raft.impl.RaftContext.lambda$runOnContext$26(RaftContext.java:334) ~[zeebe-atomix-cluster-8.4.0-SNAPSHOT.jar:8.4.0-SNAPSHOT]
	at io.atomix.utils.concurrent.SingleThreadContext$WrappedRunnable.run(SingleThreadContext.java:178) ~[zeebe-atomix-utils-8.4.0-SNAPSHOT.jar:8.4.0-SNAPSHOT]
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) ~[?:?]
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source) ~[?:?]
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source) ~[?:?]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) ~[?:?]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) ~[?:?]
	at java.base/java.lang.Thread.run(Unknown Source) [?:?]
```
Workaround: restart both the leader and the restarted pod.

Initial root cause analysis

- Server B sends join request to the leader
- Leader commits the configuration in the existing members
- Server B receives the configuration, but did not receive the commit information => configuration is kept in memory
- Server B restarts (pod was restarted)
- Server B retries join request => noop because previous join has already completed
- Leader does not sent the configuration again, because Leader assumes Server B already has it. When it sends the Append request, it results in null pointer exception.

This was not a problem before because all servers always started with the static configuration.

Proposed fixes
1. Quick fix: Leader should sent the configuration again when it receives a duplicate join request. Handle null pointer exception in the follower, by simply rejecting the AppendRequest. Eventually the follower should receive the new configuration.
2. Probably a better fix: Include the `configurationIndex` in the `AppendResponse` so that the leader can re-send the configuration if required.

### PR DESCRIPTION
**Pull request by lenaschoenburg**
Leaders track the configuration index of all members in the replication group so that it can send a configuration update to outdated members. Previously, the leader only updated the configuration index when receiving a successful configuration response. In cases where a leader initially accepted the current configuration but then loses it again, for example because of dataloss or because it never managed to commit the configuration, the leaders view of the members configuration index remained outdated: The leader thought member B had configuration index X but actually the member has no configuration at all. This could only be resolved by restarting the leader or when creating a new configuration.

As a fix, the append response now contains the configuration index. The leader updates the configuration index when it receives an append response. This allows followers to signal to the leader that they have no configuration at all or and outdated one. Before the leader tries to send the next append request, it will first send a configuration request and bring the follower up to date.

On the follower side, we now reject append requests if there is not current configuration. The append response has configuration index -1, guaranteeing that the leader will send a configure request next.

### Backwards compatibility

Adding a new field to the `AppendResponse` is safe, we have done it before: https://github.com/camunda/zeebe/pull/14063#discussion_r1311219403

Followers from the previous version do not send a configuration index. A leader from the new version will then read a 0 (default value for longs). To prevent that the follower from the previous version is then stuck in a configuration loop, we treat configuration index 0 as a special case and _don't_ update the configuration index. Effectively, we fall back to the old behavior where the configuration index is only updated when a configuration response is sent. This fallback is safe for us because we want to fix a _join_ bug where it's guaranteed that the configuration index is not 0.

Closes #15381

### CODE REVIEW

**Review by npepinpe — APPROVED** [review]
:+1: Please answer the last question, otherwise all good

File: atomix/cluster/src/test/java/io/atomix/raft/ReconfigurationTest.java [hunk 1]
```diff
@@ -206,6 +209,40 @@ void rejoinShouldBeSuccessful(@TempDir final Path tmp) {
       m3.join(id1, id2).join();
     }
 
+    @Test
+    void canJoinAgainAfterDataloss(@TempDir final Path tmp) throws IOException {
```
[the remark this answers is not in the document]
Comment (oleschoenburg) [reply atomix/cluster/src/test/java/io/atomix/raft/ReconfigurationTest.java:213]: Without this fix, this test reproduces https://github.com/camunda/zeebe/issues/15381
The leader thinks it has configured member 3 but member 3 starts again without any data and tries to join again. The leader sees member 3 already joined before and does nothing except returning success.
Without the fix, member 3 would then receive appends and run into the NPE. With this fix, member 3 notices a missing configuration, responds with an error and resets it's configuration index on the leader.

File: atomix/cluster/src/main/java/io/atomix/raft/roles/PassiveRole.java [hunk 2]
```diff
@@ -412,6 +412,11 @@ private void abortPendingSnapshots() {
   protected CompletableFuture<AppendResponse> handleAppend(final InternalAppendRequest request) {
     final CompletableFuture<AppendResponse> future = new CompletableFuture<>();
 
+    // Check that there is a configuration and reject the request if there isn't.
+    if (!checkConfiguration(request, future)) {
```
[the remark this answers is not in the document]
Comment (oleschoenburg) [reply atomix/cluster/src/main/java/io/atomix/raft/roles/PassiveRole.java:416]: Yeah I don't like it either but it matches the rest of the code as you said :shrug:

File: atomix/cluster/src/main/java/io/atomix/raft/roles/PassiveRole.java [hunk 3]
```diff
@@ -431,6 +436,16 @@ protected CompletableFuture<AppendResponse> handleAppend(final InternalAppendReq
     return future;
   }
 
+  private boolean checkConfiguration(
+      final InternalAppendRequest request, final CompletableFuture<AppendResponse> future) {
+    if (raft.getCurrentConfigurationIndex() == -1) {
+      log.debug("Rejected {}: No current configuration", request);
```
[the remark this answers is not in the document]
Comment (oleschoenburg) [reply atomix/cluster/src/main/java/io/atomix/raft/roles/PassiveRole.java:442]: It shouldn't log too much under normal circumstances but you are right that a throttled logger would be good to ensure that we don't spam the log if something goes wrong :+1:

[the remark this answers is not in the document]
Comment (oleschoenburg) [reply atomix/cluster/src/main/java/io/atomix/raft/roles/PassiveRole.java:442]: I switched to a throttled logger and changed the log type to `warn` as there is likely something wrong if this is logged too often or permanently.

File: atomix/cluster/src/main/java/io/atomix/raft/roles/LeaderAppender.java [hunk 4]
```diff
@@ -244,6 +244,24 @@ private void succeedAttempt(final RaftMemberContext member) {
...
   }
 
+  /**
+   * Updates the configuration index of the member based on the append response. This allows members
+   * which lost a configuration to update the leader and effectively request a configuration update.
+   */
+  private void updateConfigurationIndex(
+      final RaftMemberContext member, final AppendResponse response) {
+    final var configIndex = response.configurationIndex();
+    if (configIndex == 0) {
```
[the remark this answers is not in the document]
Comment (oleschoenburg) [reply atomix/cluster/src/main/java/io/atomix/raft/roles/LeaderAppender.java:254]: That said, the RollinUpdate test doesn't fail if I remove this check :disappointed: 

I'll double check, just to be sure :+1:

[the remark this answers is not in the document]
Comment (oleschoenburg) [reply atomix/cluster/src/main/java/io/atomix/raft/roles/LeaderAppender.java:254]: I verified this manually, it really is 0!

**Comment by oleschoenburg** [comment]
bors merge

**Pull request merged**